checker/src/org/checkerframework/checker/lock/LockVisitor.java

```diff
@@ -353,9 +576,319 @@ protected boolean skipContractCheck(Tree tree, FlowExpressions.Receiver expr, Fl
+                                    }
+                                }
+
+                                if (bothAreGSwithNoIndex ||
+                                  !(atypeFactory.getQualifierHierarchy().isSubtype(anno1, anno2) ||
+                                  atypeFactory.getQualifierHierarchy().isSubtype(anno2, anno1))) {
+                                // TODO: allow these strings to be localized
+
+                                String formalParam1 = null;
+
+                                if (i == 0) {
+                                    formalParam1 = "The receiver type";
+                                } else {
+                                    formalParam1 = "Parameter #" + i; // i, not i-1, so the index is 1-based
+                                }
+
+                                String formalParam2 = "parameter #" + j; // j, not j-1, so the index is 1-based
+
+                                checker.report(Result.failure(
+                                        "guardsatisfied.parameters.must.match",
+                                        formalParam1, formalParam2, invokedMethod.toString(), guardSatisfiedIndex[i], anno1, anno2), node);
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        return super.visitMethodInvocation(node, p);
+    }
+
+    // When visiting a synchronized block, issue an error if the expression
+    // has a type that implements the java.util.concurrent.locks.Lock interface.
+    // TODO: make a type declaration annotation for this rather than looking for Lock.class explicitly.
+    @Override
+    public Void visitSynchronized(SynchronizedTree node, Void p) {
+        ProcessingEnvironment processingEnvironment = checker.getProcessingEnvironment();
+
+        javax.lang.model.util.Types types = processingEnvironment.getTypeUtils();
+
+        TypeMirror lockInterfaceTypeMirror = TypesUtils.typeFromClass(types, processingEnvironment.getElementUtils(), Lock.class);
+
+        TypeMirror expressionType = types.erasure(atypeFactory.getAnnotatedType(node.getExpression()).getUnderlyingType());
+
+        if (types.isSubtype(expressionType, lockInterfaceTypeMirror)) {
+            checker.report(Result.failure(
+                    "explicit.lock.synchronized"), node);
+        }
+
+        return super.visitSynchronized(node, p);
+    }
+
+    @Override
+    public Void visitIdentifier(IdentifierTree node, Void p) {
+
+        checkAccessOfExpression(node);
+
+        return super.visitIdentifier(node, p);
+    }
+
+    @Override
+    protected void checkPreconditions(Tree tree,
+            Element invokedElement, boolean methodCall, Set<Pair<String, String>> additionalPreconditions) {
+        checkPreconditions(tree, invokedElement, methodCall, additionalPreconditions, false);
+    }
+
+    // Same contents as BaseTypeVisitor.checkPreconditions except for the addition of the
```

Comment:
I don't understand why BaseTypeVisitor.checkPreconditions wasn't modified instead of copying the code.  Why does the Lock Checker require special handling and what is that special behavior?